mars/services/task/execution/ray/context.py

```diff
@@ -12,9 +12,108 @@
+
+    def create_remote_object(self, name: str, object_cls, *args, **kwargs):
+        remote_object = object_cls(*args, **kwargs)
+        self._named_remote_objects[name] = remote_object
+
+    def destroy_remote_object(self, name: str):
+        self._named_remote_objects.pop(name, None)
+
+    async def call_remote_object(self, name: str, attr: str, *args, **kwargs):
+        remote_object = self._named_remote_objects[name]
+        meth = getattr(remote_object, attr)
+        async_meth = self._sync_to_async(meth)
+        return await async_meth(*args, **kwargs)
+
+    @staticmethod
+    @functools.lru_cache(100)
+    def _sync_to_async(func):
+        if inspect.iscoroutinefunction(func):
+            return func
+        else:
+
+            async def async_wrapper(*args, **kwargs):
+                return func(*args, **kwargs)
+
+            return async_wrapper
+
+
+class _RayRemoteObjectWrapper:
+    def __init__(self, task_state_actor: "ray.actor.ActorHandle", name: str):
+        self._task_state_actor = task_state_actor
+        self._name = name
+
+    def __getattr__(self, attr):
+        def wrap(*args, **kwargs):
+            r = self._task_state_actor.call_remote_object.remote(
+                self._name, attr, *args, **kwargs
+            )
+            return ray.get(r)
+
+        return wrap
+
+
+class _RayRemoteObjectContext:
+    def __init__(
+        self, actor_name_or_handle: Union[str, "ray.actor.ActorHandle"], *args, **kwargs
+    ):
+        super().__init__(*args, **kwargs)
+        self._actor_name_or_handle = actor_name_or_handle
+        self._task_state_actor = None
+
+    def _get_task_state_actor(self) -> "ray.actor.ActorHandle":
+        if self._task_state_actor is None:
+            if isinstance(self._actor_name_or_handle, ray.actor.ActorHandle):
+                self._task_state_actor = self._actor_name_or_handle
+            else:
+                self._task_state_actor = ray.get_actor(self._actor_name_or_handle)
+        return self._task_state_actor
+
+    @implements(Context.create_remote_object)
+    def create_remote_object(self, name: str, object_cls, *args, **kwargs):
+        task_state_actor = self._get_task_state_actor()
+        task_state_actor.create_remote_object.remote(name, object_cls, *args, **kwargs)
+        return _RayRemoteObjectWrapper(task_state_actor, name)
+
+    @implements(Context.get_remote_object)
+    def get_remote_object(self, name: str):
+        task_state_actor = self._get_task_state_actor()
+        return _RayRemoteObjectWrapper(task_state_actor, name)
+
+    @implements(Context.destroy_remote_object)
+    def destroy_remote_object(self, name: str):
+        task_state_actor = self._get_task_state_actor()
+        task_state_actor.destroy_remote_object.remote(name)
+
+
+# TODO(fyrestone): Implement more APIs for Ray.
+class RayExecutionContext(_RayRemoteObjectContext, ThreadedServiceContext):
+    """The context for tiling."""
```

Comment:
Is it possible not to inherit `ThreadedServiceContext` which has many properties and methods like `_subtask_api` that is not used in ray task executor?

Reply:
There are many APIs in the `ThreadedServiceContext` are used by the task supervisor, e.g. `get_total_n_cpu`, `get_worker_addresses`, ... So, the `RayExecutionContext` inherits from `ThreadedServiceContext`.

Maybe we can split the `ThreadedServiceContext` into two classes in the future, one for tiling and the other for executing. But, the remote object management is shared by tiling and executing. In most cases, the `create_remote_object` is for tiling, while `get_remote_object` and `destroy_remote_object` are for executing.